java/src/main/java/io/rapidpro/flows/runner/Step.java

```diff
@@ -107,11 +116,21 @@ public void addActionResult(Action.Result actionResult) {
         }
     }
 
+    public Flow getFlow() {
+        return m_flow;
+    }
+
     public List<String> getErrors() {
         return m_errors;
     }
 
     public boolean isCompleted() {
-        return m_leftOn != null;
+        return m_leftOn != null || m_terminal;
```

Comment:
But on the server they are still independent flow runs. For that flow run, the final step, as always, never has a left on. I'm not averse to changing that for everything though. If your flow run terminates, it's not wholly unreasonable to leave the last step you are on.